src/niri.rs

```diff
@@ -382,8 +382,43 @@ impl State {
         };
 
         let keyboard = self.niri.seat.get_keyboard().unwrap();
-        if keyboard.current_focus() != focus {
+        let current_focus = keyboard.current_focus();
+        if current_focus != focus {
+            if self.niri.config.borrow().input.keyboard.track_layout == TrackLayout::Window {
+                let current_layout =
+                    keyboard.with_kkb_state(self, |context| context.active_layout());
+
+                let mut new_layout = current_layout;
+                // Store the currently active layout for the surface.
```

Comment:
I guess only toplevels ever get keyboard focus, not subsurfaces and whatnot? What about popups? Those should probably follow the parent window.

Reply:
I think they grab, but not sure they receive focus? If you can test what happens it would be fine, we could resolve parents here, if it matters.

Reply:
I tested and it works with popups just fine, the layout is changed per window, so parent inheritcs.